### DISCUSSION TITLE
Will pnpm add the pnpm deduplicate command?

### DISCUSSION THREAD

**Discussion by xiaoxiaojx** [pnpm/pnpm#4143] [Q&A] [answered]
I found that when upgrading packages with the pnpm add command, there is a situation where sass: ^1.30.0 and sass: '^1.44.0' are not merged, resulting in an error 

Although it can be advertised that people use pnpm update to upgrade packages, some junior development engineers accidentally cause this problem and it is hard to find

If the pnpm deduplicate command is provided, you can force it to run once before building

**Comment by zkochan** [comment, marked as answer]
I think we can change how the add command behaves, so that it deduplicates (like update).

**Reply by xiaoxiaojx** [reply]
Thanks, this is very necessary, so that the project can be safely migrated

**Comment by anomiex** [comment]
While deduplicating on package add/install is good, I think there's also room for deduplication of existing lockfiles. Particularly when working in a monorepo using renovate, it seems easy to wind up with some workspaces' indirect deps not getting updated.

@zkochan What do you think about the idea of having a command (or option to an existing command) that works like a `pnpm install` does with no existing lockfiles, but instead of downloading package version lists like usual it instead only considers the package versions in the existing lock file? That should result in a deduplication.

**Comment by ocavue** [comment]
Since I encounter this problem frequently, I wrote my own tool to deduplicate `pnpm-lock.yaml`. It's simple but works in my projects. Feedback is welcome!

https://github.com/ocavue/pnpm-deduplicate

**Reply by gluxon** [reply]
Thank you for making this!
